blog/content/post/2021-10-05-porting_edk2_to_a13_tablet.md

```diff
@@ -0,0 +1,189 @@
+---
+title: Porting EDK II to an old Allwinner A13 tablet
+abstract: "Most ARM SoC's run U-Boot or some custom bootloader. That was the
+           case with Allwinner SoC's, until I started porting EDK II to my A13
+           tablet. In this post, I will tell you about the current UEFI support
+           status on Allwinner SoC's, my future plans, and how to test UEFI on
+           a compatible device
+          "
+cover: /img/xw711_uefi_screen.jpg
+author: artur.kowalski
+layout: post
+published: true
+date: 2021-10-05
+archives: "2021"
+
+tags:
+  - firmware
+  - edk2
+  - allwinner
+categories:
+  - Firmware
+  - Manufacturing
+
+---
+
+# Introduction
+
+I have an old cheap Allwinner A13 tablet with low memory and a weak CPU; it's
+slow even for internet browsing, so I've been using it primarily for
+development. Some time ago, I started porting EDK II to it. I plan to replace
+the current Allwinner bootloader with a full-featured UEFI implementation
+capable of booting mainline Linux, FreeBSD, and possibly Windows 10 (if I will
+find ARM32 build), all running with ACPI. Using ACPI would allow OS to boot
+without SoC specific drivers, as all basic functionality like CPU/RAM detection,
+frequency, and voltage scaling, serial port, RTC, etc., would be implemented by
+ACPI and runtime drivers.
+
+## Project status
+
+The project is still at its early stage - it does not support CPU frequency
+scaling, so it's running very slow. Currently, there is basic support for some
+hardware: UART, I2C, PMIC, SD (eMMC support is broken).
+
+It is possible to boot recent Linux with FDT (tested with v5.11).
+
+Following things need to be implemented
+
+- CPU frequency scaling
+- HDMI
+- DMA
+- UEFI variables - currently, variables are not preserved across reboots
+- ACPI
+- SD/MMC write support
+- Support for other CPUs and boards.
+
+Currently, I am working on USB peripheral mode support. It will allow making
+boot process compliant with PI specification. Currently, a small executable
```

Comment:
@pietrushnic BootROM can't load full UEFI image into RAM, so to load it I execute SPL which initializes DRAM and returns to BootROM, then I can upload the rest of UEFI.
DRAM initialization should be done from PEI and I can't go back from PEI to BootROM because that wouldn't be compliant with PI spec.

I'm not even sure whether it's possible to make boot process compliant at all - PI spec says:
> The Security (SEC) phase is the first phase in the PI Architecture architecture and is responsible for
> the following:
> • Handling all platform restart events
> • Creating a temporary memory store
> • Serving as the root of trust in the system
> • Passing handoff information to the PEI Foundation
But most ARM SoCs come with BootROM, which is the first phase.

Also BootROM has 31 KiB limit on A13 and even less on A10 and A20 (SRAM is much bigger than that). PEI may not fit. I can bypass this limit when loading over USB but not when loading from MMC, so probably I will still need some separate program just to load PEI.

I removed PI compliance from blogpost. Still I want to move DRAM initialization into PEI because doing so will allow me to fetch DRAM config from dynamic PCDs, which will be useful for overclocking.

Reply:
All the above concerns are IMO valid and should be in your blog post. It is very good question to ask "what the heck it even mean to be PI spec compliant?" - it is very hard to follow vague spec. More to that I think SEC phase was described in such way to make it as flexible as possible, because PI spec does not want to limit its adoption around platform initialization. Key point here is that PEI is bloated and system you describe has hard time with resources. Whole situation creates great opportunity for true embedded firmware development.

Now I understand why you load from USB, but I think our above discussion should in some form be introduced into the blog post.

Reply:
@pietrushnic I extended post with some information, notably, additional benefits and the problem with PEI size, still I'm thinking about PI compliance part.

Reply:
@pietrushnic I added few words about PI compliance